src/infra/dataset-update-flow-inmem/src/services/flow/flow_service_inmem.rs

```diff
@@ -0,0 +1,733 @@
+use futures::TryStreamExt;
+use kamu_core::events::DatasetEventDeleted;
+use kamu_core::{InternalError, SystemTimeSource};
+use kamu_dataset_update_flow::*;
+use kamu_task_system::*;
+use opendatafabric::{AccountID, AccountName, DatasetID};
+use tokio_stream::StreamExt;
+
+use super::active_configs_state::ActiveConfigsState;
+use super::flow_time_wheel::FlowTimeWheel;
+use super::pending_flows_state::PendingFlowsState;
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+pub struct FlowServiceInMemory {
+    state: Arc<Mutex<State>>,
+    flow_event_store: Arc<dyn FlowEventStore>,
+    time_source: Arc<dyn SystemTimeSource>,
+    task_scheduler: Arc<dyn TaskScheduler>,
+    flow_configuration_service: Arc<dyn FlowConfigurationService>,
+    dependency_graph_service: Arc<dyn DependencyGraphService>,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+struct State {
+    active_configs: ActiveConfigsState,
+    pending_flows: PendingFlowsState,
+    time_wheel: FlowTimeWheel,
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[component(pub)]
+#[interface(dyn FlowService)]
+#[interface(dyn AsyncEventHandler<TaskEventFinished>)]
+#[interface(dyn AsyncEventHandler<DatasetEventDeleted>)]
+#[interface(dyn AsyncEventHandler<FlowConfigurationEventModified>)]
+#[scope(Singleton)]
+impl FlowServiceInMemory {
+    pub fn new(
+        flow_event_store: Arc<dyn FlowEventStore>,
+        time_source: Arc<dyn SystemTimeSource>,
+        task_scheduler: Arc<dyn TaskScheduler>,
+        flow_configuration_service: Arc<dyn FlowConfigurationService>,
+        dependency_graph_service: Arc<dyn DependencyGraphService>,
+    ) -> Self {
+        Self {
+            state: Arc::new(Mutex::new(State::default())),
+            flow_event_store,
+            time_source,
+            task_scheduler,
+            flow_configuration_service,
+            dependency_graph_service,
+        }
+    }
+
+    async fn run_current_timeslot(&self) {
+        let planned_flows: Vec<_> = {
+            let mut state = self.state.lock().unwrap();
+            state.time_wheel.take_nearest_planned_flows()
+        };
+
+        let planned_task_futures: Vec<_> = planned_flows
+            .iter()
+            .map(async move |flow_id| {
+                let mut flow = Flow::load(*flow_id, self.flow_event_store.as_ref())
+                    .await
+                    .int_err()?;
+
+                self.schedule_flow_task(&mut flow).await?;
+                Ok(())
+            })
+            .collect();
+
+        let results = futures::future::join_all(planned_task_futures).await;
+        results
+            .into_iter()
+            .filter(|res| res.is_err())
+            .map(|e| e.err().unwrap())
+            .for_each(|e: InternalError| {
+                tracing::error!("Scheduling flow failed: {:?}", e);
```

Comment:
Don't format values into the message - use structured fields instead:

```rust
tracing::error!(error = ?e, "Scheduling flow failed");
```

Reply:
We actually should have a standard of how to capture error in tracing correctly. Here's what I was doing in HTTP adapter:
```rust
tracing::error!(
                error = ?self.source,
                error_msg = %self.source,
                status_code = %self.status_code,
                "Internal API error",
);
```

This captures both the debug representation and the display message, but we'll likely want a backtrace too.

Reply:
I've corrected this one, but I haven't really started working yet on tracing in these new crates.